llmfoundry/models/mpt/modeling_mpt.py

```diff
@@ -92,6 +116,33 @@ def __init__(self, config: MPTConfig):
                 **config.to_dict(),
             ) for _ in range(config.n_layers)
         ])
+        if config.tensor_parallel_mlp and is_torch_2_or_higher():
+            if enable_2d_with_fsdp():
+                device_type = 'cuda' if get_device(
+                    None).name == 'gpu' else 'cpu'
+                world_size = dist.get_world_size()
+                node_count = world_size // dist.get_local_world_size()
+                twod_mesh = DeviceMesh(device_type=device_type,
+                                       mesh=torch.arange(0, world_size).view(
+                                           node_count, -1))
+                new_blocks = nn.ModuleList()
+                for idx, block in enumerate(self.blocks):
+                    block = parallelize_module(
+                        module=block,
+                        device_mesh=twod_mesh,
+                        parallelize_plan={'ffn': PairwiseSequenceParallel()},
+                        tp_mesh_dim=1)
+                    # Call to parallelize_module moves params to gpu if they are cpu params.
+                    # Move them back to cpu so that FSDP wrapping sees all params on cpu.
+                    # Othewise FSDP wrapping fails as it sees some params on cpu and others on gpu.
+                    if config.init_device == 'cpu':
+                        block = block.to('cpu')
+                    new_blocks.append(block)
+                self.blocks = new_blocks
+            else:
+                raise ValueError('Enabling 2D parallelism with FSDP failed (call to enable_2d_with_fsdp). ' + \
+                    'Please make sure your torch installation supports 2D parallelism with FSDP')
+
```

Comment:
Could this be packaged away into a util fn so that we can keep the modeling_mpt.py file clean?